### ISSUE TITLE
Possible bug: Complex named function params with defaults and rest causes errors.

### ISSUE THREAD

**Issue by dsifford** [Microsoft/TypeScript#14026]
**TypeScript Version:**  `2.1.1` & `rc`

Sorry if this has already been reported. I searched around and couldn't find exactly this.

This is sort of difficult to describe, so here's an example:

```ts
interface Foo {
    bar?: any;
    baz?: any;
}

function foobar({ bar = {}, ...opts }: Foo = {}) {
    console.log(bar);
    console.log(opts.baz);
}

foobar(); // Throws error
foobar({ baz: 'hello' }); // No error
foobar({ bar: { greeting: 'hello' } }); // No error
```

[Playground example](http://www.typescriptlang.org/play/index.html#src=interface%20Foo%20%7B%0D%0A%20%20%20%20bar%3F%3A%20any%3B%0D%0A%20%20%20%20baz%3F%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Afunction%20foobar(%7B%20bar%20%3D%20%7B%7D%2C%20...opts%20%7D%3A%20Foo%20%3D%20%7B%7D)%20%7B%0D%0A%20%20%20%20console.log(bar)%3B%0D%0A%20%20%20%20console.log(opts.baz)%3B%0D%0A%7D%0D%0A%0D%0Afoobar()%3B%20%2F%2F%20Throws%20error%0D%0Afoobar(%7B%20baz%3A%20'hello'%20%7D)%3B%20%2F%2F%20No%20error%0D%0Afoobar(%7B%20bar%3A%20%7B%20greeting%3A%20'hello'%20%7D%20%7D)%3B%20%2F%2F%20No%20error)

**Expected behavior:**
I'd expect `bar` to still be initialized and available in the function call as an empty object when no args are given to the function.

**Actual behavior:**

The following error gets thrown (reproducible in the playground example above).
```
Uncaught TypeError: Cannot read property 'bar' of undefined
    at foobar (<anonymous>:11:16)
    at <anonymous>:16:1
    at HTMLButtonElement.excuteButton.onclick (http://www.typescriptlang.org/play/playground.js:242)
```

**Comment by RyanCavanaugh** [comment]
We should be emitting the parameter default above the destructuring logic, not below it

**Comment by sandersn** [comment]
Fix is up at #14090
